### DISCUSSION TITLE
CSRF gives Error 419 due to inactivity even when I’m logged out

### DISCUSSION THREAD

**Discussion by elfeffe** [livewire/livewire#1031] [4. General]
Is there any way to prevent it?
In a eCommerce app this is weird, a visitor leaves the page in a category and when returning it shows that error. 
Is this because livewire uses POST requests?
You can check it at novapackages.com

**Comment by mblarsen** [comment]
See this issue #537 for a possible workaround.

**Reply by elfeffe** [reply]
I will try it, thank you

**Comment by calebporzio** [comment]
Agreed there should be a more official solution surrounding this issue.

What would you expect to happen? Or want to happen?

**Reply by elfeffe** [reply]
This extenson has fixd the issue
https://github.com/GeneaLabs/laravel-caffeine 

I think a solution will be to create a component that just refreshes CSRF code every X minutes, And then replace CSRF tag on HTML

**Reply by mblarsen** [reply]
Or a less aggressive approach would be to leave add hook on window.livewire to let the user define deal with it. E.g. refresh the page, show a message, hook into above mentioned caffeine or whatever the dev sees fit.

**Reply by elfeffe** [reply]
> Agreed there should be a more official solution surrounding this issue.
> 
> What would you expect to happen? Or want to happen?

I think I have never answered. What I expect is my e-commerce non logged visitors can continue using my site after they leave the computer with one of my products on their screen. 
I think a simple solution is to switch the csrf code with a livewire component that auto refresh every x minutes with the actual code. 
Laravel-caffeine seems to be doing the job.

**Reply by booni3** [reply]
Caffeine is a potential security risk as it keeps the user's session alive indefinitely i.e. they will never be logged out. This defeats the point of the "remember me" checkbox.

On a traditional site, if your session expired, you would be redirected to login the next time you tried to click an auth route.

The issue with Livewire currently is that the 419 error is hidden in the network tab and unless you change the default behaviour the user would probably think the site is broken.

Ideally, I think you need to either:
- Redirect user to log in after session timeout when any request is sent out of session. This can easily be done in the handler but unfortunately, you loose your request. In a traditional app your request would be re-sent after you logged in. In some ways this is maybe not possible with an SPA type behavior.

- Redirect user to login as soon as they bring the page back into focus. This is probably better UX because they will have not spent time filling a form or trying to link to somewhere else first. I believe this is how the Laravel apps such as Vapor handle the session expiry.

**Reply by phamquocbuu** [reply]
I've tried to create a Middleware which handled before everything in `protected $middleware`. But a Livewire request expects a `text/json` response, not `text/javascript` nor `text/html`. And a redirect fired from a middleware do nothing with a Livewire request, moreover it cause an error that nothing more in Js will run.

However, this can handle well for the login form, which is opened and stayed after the session expired.

```
<?php

namespace App\Http\Middleware;

use Closure;
use Illuminate\Support\Facades\Auth;
use Illuminate\Support\Facades\Session;
use Illuminate\Support\Str;

class Check419 {
    /**
     * Handle an incoming request.
     *
     * @param \Illuminate\Http\Request $request
     * @param \Closure $next
     * @return mixed
     */
    public function handle($request, Closure $next) {
        $response = $next($request);

        if ($response->status() === 419) {
            $acceptable = $request->getAcceptableContentTypes();
            if (isset($acceptable[0]) && Str::contains($acceptable[0], ['/html', '+xml'])) {
                return $response;
            }

            Auth::logout();
            Session::flush();
            return redirect()->to('/login');
        }

        return $response;
    }
}
```

```
<?php
use Illuminate\Foundation\Http\Kernel as HttpKernel;

class Kernel extends HttpKernel
{
    /**
     * The application's global HTTP middleware stack.ssss
     * The application's global HTTP middleware stack.
     *
     * These middleware are run during every request to your application.
     *
     * @var array
     */
    protected $middleware = [
        \App\Http\Middleware\Check419::class,
// ...
```
